### PR TITLE
Closes #2415 replaced hard coding of expr(USUBJID, STUDYID) or expr(USUBJID) with …

### LINKED ISSUE #2415: General Issue: Make subject keys flexible in all functions
### Background Information

@bundfussr did a review of the admiral functions and found 2 occurrences where `STUDYID` and `USUBJID` are hardcoded instead of using `get_admiral_option("subject_keys")` as per https://pharmaverse.github.io/admiral/reference/#admiral-options.

These are `derive_vars_atc()` and `create_single_dose_dataset()`, for example:

```
derive_vars_atc <- function(dataset,
                            dataset_facm,
                            by_vars = exprs(USUBJID, CMREFID = FAREFID),
                            value_var = FASTRESC)
```

### Definition of Done

These functions should be flexible so as if ever `set_admiral_options(subject_keys = exprs(STUDYID, USUBJID2))` was set by user then these variables would be used everywhere instead of `USUBJID`.

### PR DESCRIPTION
**Pull request by ProfessorP-beep**
…!!!get_admiral_option("subject_keys") to allow for flexibility in using the function. There is an error in the bds_exposure.Rmd markdown on line 361, chunk 22 that needs to be fixed.

Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Review the [Cheat Sheet](https://github.com/pharmaverse/admiral/blob/main/inst/cheatsheet/admiral_cheatsheet.pdf). Make any required updates to it by editing the file `inst/cheatsheet/admiral_cheatsheet.pptx` and re-upload a PDF version of it to the same folder.
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers). A Developer Notes section is available in `NEWS.md` for tracking developer-facing issues.  
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by bms63** [comment]
Looking nice @ProfessorP-beep!  Can you just run styler on your code

**Comment by ProfessorP-beep** [comment]
I'm still getting the error with pkgdown::build_site(), but everything else, including running the .Rmd checks out. It's running again right now after merging an update.

**Comment by bms63** [comment]
> I'm still getting the error with pkgdown::build_site(), but everything else, including running the .Rmd checks out. It's running again right now after merging an update.

I will see if I can build site on my end

**Comment by bms63** [comment]
I was able to build website - I will review PR.  @manciniedoardo can you also help with review please.

**Review by manciniedoardo — CHANGES_REQUESTED** [review]
Left some comments, thanks @ProfessorP-beep !

**Comment by manciniedoardo** [comment]
@ProfessorP-beep sorry, would it be possible to please update the occasions where we have `exprs(get_admiral_option("subject_keys")` to just say `get_admiral_option("subject_keys")`? I believe that would be the right syntax. thanks

**Comment by ProfessorP-beep** [comment]
@manciniedoardo Already done. I realized my mistake after that last push. New one coming as soon as the checks are done.

**Comment by ProfessorP-beep** [comment]
Sorry, hold off on reviewing on minute. Some of the changes weren't pushed through.

**Comment by ProfessorP-beep** [comment]
Hey @manciniedoardo, sorry about that. The latest pull is ready for review.

**Review by bms63 — COMMENTED** [review]
@manciniedoardo can you finish reviewing this one please

I'm still a bit weary of this update.

@rossfarrugia and @bundfussr you find this more helpful then the original? Like folks in Roche are cool with this and pick it up fine?  My little fingers hurt typing this all out

image

**Comment by rossfarrugia** [comment]
@bms63 any user could hardcode the subject identifiers instead if they were only ever going to use `STUDYID`, `USBUJID` here, but for our package to follow our own instructions at https://pharmaverse.github.io/admiral/reference/#admiral-options we have to follow the more generic way, otherwise an option that we ourselves advertise for users doesn't actually consistently work across the package.

File: inst/templates/ad_adae.R
Comment (bundfussr) [reply inst/templates/ad_adae.R]: > Also, I believe in this case you could do `by_vars = get_admiral_option("subject_keys")` directly rather than quoting and unquoting `by_vars = exprs(!!!get_admiral_option("subject_keys"))` (this is applicable in lots of other places as well.) You'd only need to use `exprs` and unquoting when you want to use other by vars as well, for instance `by_vars = exprs(!!!get_admiral_option("subject_keys"), AEDECOD)` - @bundfussr do you agree?

I agree.

**Comment by bms63** [comment]
> @bms63 any user could hardcode the subject identifiers instead if they were only ever going to use `STUDYID`, `USBUJID` here, but for our package to follow our own instructions at https://pharmaverse.github.io/admiral/reference/#admiral-options we have to follow the more generic way, otherwise an option that we ourselves advertise for users doesn't actually consistently work across the package.

Okay!  I understand how options work :)  I feel like I'm not making myself very clear with my concerns...so maybe my concerns for the users is moot! :)   If you all think this update makes sense for the users, then full steam ahead!!

**Comment by bms63** [comment]
and thanks @ProfessorP-beep for seeing this through!!  This is how it works on the team :) We do PRs and I get a little huffy and puffy ask questions funny questions - which usually are me being paranoid.

**Comment by rossfarrugia** [comment]
Thanks @bms63 - all fair questions, but i guess these all would have been discussed when we first added this option. Now that we have it offered in the package, we just need to make sure it works :)

File: vignettes/generic.Rmd
Comment (bundfussr) [inline vignettes/generic.Rmd:222]: This looks confusing to me. We don't have any rules or convention on the number of elements in "subject_keys" or which variable is placed where. You could have `exprs(USUBJID, STUDYID)`, `exprs(STUDYID, SITEID, SUBJID)`, or `exprs(USUBJID)`. I.e., selecting a certain element is not a good idea because in general you don't know what you get.

I would either add `STUDYID` to the data and use `get_admiral_option("subject_keys")` instead of `exprs(!!get_admiral_option("subject_keys")[[2]])` or just leave it as it is.

@bms63 , @rossfarrugia , what do you think?

**Review by bundfussr — COMMENTED** [review]
There are a few missed occurrences:
```
(base) bundfuss@rstudio:(templates)> grep USUBJID *.R
ad_adlbhy.R:  select(STUDYID, USUBJID, TRT01A, PARAMCD, LBSEQ, ADT, AVISIT, ADY, AVAL, ANRHI, CRIT1, CRIT1FL)
ad_adlbhy.R:  select(STUDYID, USUBJID, TRT01A) %>% # add AVISIT, ADT for by visit
ad_adlbhy.R:  select(STUDYID, USUBJID, TRT01A, CRIT1FL, BILI_CRITFL) %>% # add AVISIT, ADT for by visit
ad_advs.R:  #   by_vars = exprs(STUDYID, USUBJID, PARAMCD, BASETYPE),
```

File: vignettes/generic.Rmd
Comment (rossfarrugia) [reply vignettes/generic.Rmd:222]: add `STUDYID` to the data and use `get_admiral_option("subject_keys")` would be my preference. i don't like the `[[2]]` option either - as risky for future robustness.

Comment (bms63) [reply vignettes/generic.Rmd:222]: This is being used in `derive_var_trtemfl()` as well.  I flagged as concerning for me https://github.com/pharmaverse/admiral/pull/2362#issuecomment-2144118821

File: inst/templates/ad_adex.R
Comment (ProfessorP-beep) [reply inst/templates/ad_adex.R:162]: Gotcha, I'll take a look at these. I think I was getting some error messages when using this with select(), but I'll look again and post it here if it comes up.

**Comment by ProfessorP-beep** [comment]
> and thanks @ProfessorP-beep for seeing this through!! This is how it works on the team :) We do PRs and I get a little huffy and puffy and ask funny questions - which usually are me being paranoid.

No problem! I'm also learning a lot more so this is cool. I'm looking through the discussion now.

File: vignettes/generic.Rmd
Comment (bms63) [reply vignettes/generic.Rmd:222]: Hi @ProfessorP-beep do you have time to make these final changes today?

Comment (ProfessorP-beep) [reply vignettes/generic.Rmd:222]: Yea, I should be able to get to it this afternoon / early evening.

File: inst/templates/ad_adeg.R
Comment (bms63) [reply inst/templates/ad_adeg.R:168]: @ProfessorP-beep did this one need an update?

Comment (ProfessorP-beep) [reply inst/templates/ad_adeg.R:168]: Oops looks like I missed it. Can update in a bit.

File: inst/templates/ad_adlbhy.R
Comment (rossfarrugia) [inline inst/templates/ad_adlbhy.R:46]: why was this one left behind with just the order of USUBJID and STUDYID swapped? noticed this by chance as was using this PR to prompt same changes in admiralonco. please double check any others

Comment (rossfarrugia) [reply inst/templates/ad_adlbhy.R:46]: as FYI should be `select(!!!get_admiral_option("subject_keys"), TRT01A, ...etc)`

**Comment by bms63** [comment]
@bundfussr @rossfarrugia @manciniedoardo - This update isn't critical to me for 1.1.  Can we implement for 1.2 as it is actually quite large and seems to be opening lots of questions?

**Comment by manciniedoardo** [comment]
> @bundfussr @rossfarrugia @manciniedoardo - This update isn't critical to me for 1.1. Can we implement for 1.2 as it is actually quite large and seems to be opening lots of questions?

Ok, let's leave until 1.2 to ensure we do this right and don't have to come back to it later. I still think it's an important update to make, and we can definitely still leverage all the work done by @ProfessorP-beep in this PR, but what we have now is not _wrong_ per se, just a bit inconsistent.

**Comment by bms63** [comment]
Hi @ProfessorP-beep, we can talk about this tomorrow, but going to pause on this PR until after we ship 1.1 release.  This will be our first PR into 1.2!

**Comment by ProfessorP-beep** [comment]
> Hi @ProfessorP-beep, we can talk about this tomorrow, but going to pause on this PR until after we ship 1.1 release. This will be our first PR into 1.2!

Sounds good, I'm just gonna push the additional changes I've done since then. There is probably still some review to be done to make sure all cases are addressed.

**Pull request closed without merging**
